core/src/sigverify_shreds.rs

```diff
@@ -0,0 +1,395 @@
+                                std::u64::MAX
+                            }
+                        })
+                        .collect()
+                })
+                .collect()
+        })
+    });
+    let mut keys: HashMap<Pubkey, Vec<u64>> = HashMap::new();
+    for batch in slots.iter() {
+        for slot in batch.iter() {
+            let key = slot_leaders.get(slot).unwrap();
+            keys.entry(*key).or_insert(vec![]).push(*slot);
+        }
+    }
+    let mut pubkeys: PinnedVec<Pubkey> = recycler_pubkeys.allocate("shred_gpu_pubkeys");
+    let mut slot_to_key_ix = HashMap::new();
+    for (i, (k, slots)) in keys.iter().enumerate() {
+        pubkeys.push(*k);
+        for s in slots {
+            slot_to_key_ix.insert(s, i);
+        }
+    }
+    let mut offsets = recycler_offsets.allocate("shred_offsets");
+    slots.iter().for_each(|packet_slots| {
+        packet_slots.iter().for_each(|slot| {
+            offsets.push((slot_to_key_ix.get(slot).unwrap() * size_of::<Pubkey>()) as u32);
+        });
+    });
+    //HACK: Pubkeys vector is passed along as a `Packets` buffer to the GPU
+    //TODO: GPU needs a more opaque interface, which can handle variable sized structures for data
+    //Pad the Pubkeys buffer such that it is bigger than a buffer of Packet sized elems
+    let num_in_packets =
+        (pubkeys.len() * size_of::<Pubkey>() + (size_of::<Packet>() - 1)) / size_of::<Packet>();
+    trace!("num_in_packets {}", num_in_packets);
+    //number of bytes missing
+    let missing = num_in_packets * size_of::<Packet>() - pubkeys.len() * size_of::<Pubkey>();
+    trace!("missing {}", missing);
+    //extra Pubkeys needed to fill the buffer
+    let extra = (missing + size_of::<Pubkey>() - 1) / size_of::<Pubkey>();
+    trace!("extra {}", extra);
+    trace!("pubkeys {}", pubkeys.len());
+    for _ in 0..extra {
+        pubkeys.push(Pubkey::default());
+        trace!("pubkeys {}", pubkeys.len());
+    }
+    trace!("pubkeys {:?}", pubkeys);
+    trace!("offsets {:?}", offsets);
+    (pubkeys, offsets)
+}
+
+fn shred_gpu_offsets(
+    mut pubkeys_end: u32,
+    batches: &[Packets],
+    recycler_offsets: &Recycler<TxOffset>,
+) -> (TxOffset, TxOffset, TxOffset, Vec<Vec<u32>>) {
+    let mut signature_offsets = recycler_offsets.allocate("shred_signatures");
+    let mut msg_start_offsets = recycler_offsets.allocate("shred_msg_starts");
+    let mut msg_sizes = recycler_offsets.allocate("shred_msg_sizes");
+    let mut v_sig_lens = vec![];
+    for batch in batches {
+        let mut sig_lens = Vec::new();
+        for packet in &batch.packets {
+            let sig_start = pubkeys_end;
+            let sig_end = sig_start + size_of::<Signature>() as u32;
+            let msg_start = sig_end;
+            let msg_end = sig_start + packet.meta.size as u32;
+            signature_offsets.push(sig_start);
+            msg_start_offsets.push(msg_start);
+            msg_sizes.push(msg_end - msg_start);
```

Comment:
I might be wrong. But, I think this could be very huge, wrapping across `0` when packet.meta.size is maliciously smaller than `size_of::<Signature>()` because there is no bounds checking? This will results in a DOS?

Reply:
the packet memory is allocated regardless of the meta size.

Reply:
@sakridge does the GPU handle sig_lens(0) for a packet?  like the packet has no signatures?

Reply:
i would want it to fail the packet

Reply:
> the packet memory is allocated regardless of the meta size.

I see. Anyway, proper bounds checks are in place now, this is not a problem anymore.

Reply:
(As far as I read [the code](https://github.com/solana-labs/solana-perf-libs/blob/master/src/cuda-ecc-ed25519/verify.cu) corectly, GPU doesn't handle sig_lens(0) because basically GPU only sees a flattened list of [sig, blob] and has no notion of `packets`. So in other words, it cannot be aware of messages with no sig. That's why we [handle the case at the CPU side](https://github.com/solana-labs/solana/pull/6520/files#diff-ad695894f5a727f4ce60e68987c4e35aR279).)